unittest/freedv_700d_comprx.c

```diff
@@ -0,0 +1,126 @@
+/*---------------------------------------------------------------------------*\
+
+  FILE........: freedv_700d_comprx.c
+  AUTHOR......: David Rowe
+  DATE CREATED: July 2022
+
+  Complex valued rx to support ctests.  Includes a few operations that will
+  only work if complex Tx and Rx signals are being handled correctly.
+ 
+\*---------------------------------------------------------------------------*/
+
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "freedv_api.h"
+#include "freedv_api_internal.h"
+#include "ofdm_internal.h"
+#include "codec2_cohpsk.h"
+#include "comp_prim.h"
+
+int main(int argc, char *argv[]) {
+
+    /* with no arguments then run with no test code */
+    int test_num = 0;
+    if (argc == 2) {
+        if (strcmp(argv[1],"tx") == 0) {
+            test_num = 1;
+        }
+        if (strcmp(argv[1],"rx") == 0) {
+            test_num = 2;
+        }
+    }
+    fprintf(stderr,"%d\n", test_num);
+    
+    struct freedv *freedv;
+    freedv = freedv_open(FREEDV_MODE_700D);
+    assert(freedv != NULL);
+
+    /* note API functions to tell us how big our buffers need to be */
+    short speech_out[freedv_get_n_max_speech_samples(freedv)];
+    short demod_in[2*freedv_get_n_max_modem_samples(freedv)];
+    COMP  demod_in_comp[2*freedv_get_n_max_modem_samples(freedv)];
+    
+    /* set up small freq offset */
+    float foff_hz = 25;
+    COMP phase_ch; phase_ch.real = 1.0; phase_ch.imag = 0.0;
+
+    /* set complex sine wave interferer at -fc */
+    COMP interferer_phase = {1.0,0.0};
+    COMP interferer_freq;
+    interferer_freq.real = cos(2.0*M_PI*freedv->ofdm->tx_centre/FREEDV_FS_8000);
+    interferer_freq.imag = sin(2.0*M_PI*freedv->ofdm->tx_centre/FREEDV_FS_8000);
+    interferer_freq = cconj(interferer_freq);
+
+    /* log a file of demod input samples for plotting in Octave */
+    FILE *fdemod = fopen("demod.f32","wb"); assert(fdemod != NULL);
+
+    /* measure demod input power, interferer input power */
+    float power_d = 0.0; float power_interferer = 0.0; 
+
+    int frames = 0, sum_sync = 0, frames_snr = 0; float sum_snr = 0.0;
+    size_t nin, nout;
+    nin = freedv_nin(freedv);
+
+    while(fread(demod_in, sizeof(short), 2*nin, stdin) == 2*nin) {
+        for(int i=0; i<nin; i++) {
+            demod_in_comp[i].real = (float)demod_in[2*i];
+            demod_in_comp[i].imag = (float)demod_in[2*i+1];
+            //demod_in_comp[i].imag = 0;
```

Comment:
Will the commented line be useful for any future testing?

Reply:
Yes. I found it pretty useful when developing this test (ie to test the design of the test), and this line is discussed in the comment below so I wanted to make it explicit.  Another possibility would be a command line option, and even a ctest to test the test fails when it should.

Reply:
OK, we can keep in here for now.